src/cache.rs

```diff
@@ -226,33 +307,104 @@ impl<T: Hash + PartialEq> Constraint<T> {
     /// Enter a constraint for a call to an immutable function.
     #[inline]
     pub fn push(&self, args: T, ret: u128, mutable: bool) {
-        let both = hash(&(&args, ret));
-        self.push_inner(Call { args, ret, both, mutable });
+        let args_hash = hash(&args);
+        let both = hash(&(args_hash, ret));
+        self.0.write().push_inner(Cow::Owned(Call {
+            args,
+            args_hash,
+            ret,
+            both,
+            mutable,
+        }));
+    }
+
+    /// Whether the method satisfies as all input-output pairs.
+    #[inline]
+    pub fn validate<F>(&self, mut f: F) -> bool
+    where
+        F: FnMut(&T) -> u128,
+    {
+        self.0.read().calls.iter().all(|entry| f(&entry.args) == entry.ret)
+    }
+
+    /// Whether the method satisfies as all input-output pairs.
+    #[inline]
+    pub fn validate_with_id<F>(&self, mut f: F, id: usize) -> bool
+    where
+        F: FnMut(&T) -> u128,
+    {
+        let accelerator = accelerator(id);
+        let inner = self.0.read();
+        if let Some(accelerator) = accelerator {
+            let mut map = accelerator.lock();
+            inner.calls.iter().all(|entry| {
+                *map.entry(entry.both).or_insert_with(|| f(&entry.args)) == entry.ret
+            })
+        } else {
+            inner.calls.iter().all(|entry| f(&entry.args) == entry.ret)
+        }
+    }
+
+    /// Replay all input-output pairs.
+    #[inline]
+    pub fn replay<F>(&self, mut f: F)
+    where
+        F: FnMut(&T),
+    {
+        self.0
+            .read()
+            .calls
+            .iter()
+            .filter(|call| call.mutable)
+            .for_each(|call| {
+                f(&call.args);
+            });
```

Comment:
please avoid for_each here, too

Reply:
I'm not so sure I agree that it's more readable, and in my experience `for_each` led to better optimizations, but I changed it to match your preferred code style.